File: 438/find-all-anagrams-in-a-string.java
class Solution {
    int [] H1;
    int [] H2;
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> ans = new ArrayList<Integer>();
        int N1 = s.length(), N2 = p.length();
        if(N1<N2) return ans;

        H1 = new int[26];
        H2 = new int[26];

        for(int i=0; i<N2; i++){//先前面n2位，讓長度與p相同
            H1[s.charAt(i)-'a']++;
        }
        for(int i=0; i<N2; i++){
            H2[p.charAt(i)-'a']++;
        }
        if(check()) ans.add(0);

        for(int i=1; i<=N1-N2; i++){
            H1[s.charAt(i+N2-1)-'a']++;
            H1[s.charAt(i-1)-'a']--;
            if(check()) ans.add(i);
        }
        return ans;
    }
    boolean check(){
        for(int i=0; i<26; i++){
            if(H1[i]!=H2[i]) return false;
        }
        return true;
    }
}

File: 438/find-all-anagrams-in-a-string.py
# s 比較長, p 比較短, p重新組合的anagram, 出現在 s 的哪些開始地方
# 使用 running window 逐一測試即可
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        sN, pN = len(s), len(p) # 對應的長度
        if sN < pN: return [] # s長度不夠, 沒辧法組出 anagram

        pH = Counter(p) # 用來統計p的字母出現次數
        sH = Counter(s[:pN]) # 用來統計s的前pN個數字
        ans = []
        print(pH-sH)
        if len(pH-sH)==0: ans.append(0) # 第0個比較
        for i in range(sN-pN): # 長度差,便是迴圈要做的事
            sH[s[i]] -= 1
            sH[s[i+pN]] += 1
            if len(pH-sH)==0: ans.append(i+1)
        return ans
            
